csrc/welford.cu

```diff
@@ -61,25 +64,27 @@ __device__ __forceinline__ T reduce_block(T *x, T val)
 {
   int tid = threadIdx.y*blockDim.x + threadIdx.x;
   int blockSize = blockDim.x * blockDim.y;
+  int lane = tid % WARP_SIZE;
+  int wid = tid / WARP_SIZE;
 
-  if (blockSize > 32) {
+  if (blockSize > WARP_SIZE) {
     val = warp_reduce_sum(val);
-    if (tid % WARP_SIZE == 0)
-      x[tid/WARP_SIZE] = val;
+    if (lane == 0)
+      x[wid] = val;
 
     __syncthreads();
 
-    val = (tid < blockSize / WARP_SIZE? x[tid%WARP_SIZE] : T(0));
+    val = (tid < blockSize / WARP_SIZE? x[lane] : T(0));
   }
 
-  if(tid/WARP_SIZE==0) val = warp_reduce_sum(val);
+  if(wid==0) val = warp_reduce_sum(val);
 
   return val;
 }
 
 #define ELEMENTS_PER_ITER 4 // enables concurrency within each thread to hide latency
 #define ELEMENTS_PER_THREAD 16
-#define OPTIMAL_TILE_W 32
+#define OPTIMAL_TILE_W WARP_SIZE
 #define MAX_H_BLOCK 128
 #define MAX_BLOCK_SIZE 512
```

Comment:
@jeffdaily Apart from the warp size fix, should we also check the launch_bounds?  Here we have MAX_BLOCK_SIZE set to 512.

Reply:
Still need to figure this out.  If I do not explicitly use `__launch_bounds__`, (so default must be 256) unit tests pass.  If I specify 512, unit tests fail.  If I specify 1024, unit tests pass again.

Reply:
Was not able to reproduce with latest pushed commit.